Index: mono/mcs/class/corlib/System.Resources/ResourceManager.cs
===================================================================
--- mono/mcs/class/corlib/System.Resources/ResourceManager.cs.orig	2010-11-09 22:08:15.000000000 -0700
+++ mono/mcs/class/corlib/System.Resources/ResourceManager.cs	2010-11-09 22:08:52.000000000 -0700
@@ -328,10 +328,14 @@
 
 			ResourceSet set;
 			
+#if false
 			/* if we already have this resource set, return it */
 			set = (ResourceSet) ResourceSets [culture];
 			if (set != null)
 				return set;
+#else
+			set = null;
+#endif
 
 			if (NonExistent.Contains (culture))
 				return null;
