Index: mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs
===================================================================
--- mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs	(revision 152577)
+++ mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs	(working copy)
@@ -5142,6 +5137,18 @@
 			UpdateBounds(x, y, width, height, width - (rect.Right - rect.Left), height - (rect.Bottom - rect.Top));
 		}
 
+		internal static bool m_ignoreUpdates = false;
+
+		public static void IgnoreUpdates()
+		{
+			m_ignoreUpdates = true;
+		}
+
+		public static void UnignoreUpdates()
+		{
+			m_ignoreUpdates = false;
+		}
+
 		[EditorBrowsable(EditorBrowsableState.Advanced)]
 		protected void UpdateBounds(int x, int y, int width, int height, int clientWidth, int clientHeight) {
 			// UpdateBounds only seems to set our sizes and fire events but not update the GUI window to match
@@ -5169,20 +5176,24 @@
 			client_size.Width=clientWidth;
 			client_size.Height=clientHeight;
 
-			if (moved) {
-				OnLocationChanged(EventArgs.Empty);
+			if (!m_ignoreUpdates)
+			{
+				if (moved) {
 
-				if (!background_color.IsEmpty && background_color.A < byte.MaxValue)
-					Invalidate ();
-			}
+					OnLocationChanged(EventArgs.Empty);
 
-			if (resized) {
-				OnSizeInitializedOrChanged ();
-				OnSizeChanged(EventArgs.Empty);
-#if NET_2_0
-				OnClientSizeChanged (EventArgs.Empty);
-#endif
+					if (!background_color.IsEmpty && background_color.A < byte.MaxValue)
+						Invalidate ();
+				}
+
+				if (resized) {
+					OnSizeInitializedOrChanged ();
+					OnSizeChanged(EventArgs.Empty);
+	#if NET_2_0
+					OnClientSizeChanged (EventArgs.Empty);
+	#endif
 			}
+			}
 		}
 
 		[EditorBrowsable(EditorBrowsableState.Advanced)]
