Index: mono/mcs/class/System.Configuration/System.Configuration/InternalConfigurationHost.cs
===================================================================
--- mono/mcs/class/System.Configuration/System.Configuration/InternalConfigurationHost.cs	(revision 132913)
+++ mono/mcs/class/System.Configuration/System.Configuration/InternalConfigurationHost.cs	(working copy)
@@ -71,6 +71,8 @@
 		public virtual Type GetConfigType (string typeName, bool throwOnError)
 		{
 			Type type = Type.GetType (typeName);
+			if (type == null && typeName.Split(',').Length == 1) // try System.dll if nothing else specified
+				type = Type.GetType (typeName + ",System");
 			if (type == null && throwOnError)
 				throw new ConfigurationErrorsException ("Type '" + typeName + "' not found.");
 			return type;
Index: mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridViewRowCollection.cs
===================================================================
--- mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridViewRowCollection.cs	(revision 132913)
+++ mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridViewRowCollection.cs	(working copy)
@@ -252,6 +252,8 @@
 
 		public virtual void Clear ()
 		{
+			dataGridView.CurrentCell = null;
+
 			int total = list.Count;
 			
 			for (int i = 0; i < total; i++) {
Index: mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridViewColumnCollection.cs
===================================================================
--- mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridViewColumnCollection.cs	(revision 132913)
+++ mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridViewColumnCollection.cs	(working copy)
@@ -101,6 +101,8 @@
 
 		public virtual void Clear ()
 		{
+			dataGridView.CurrentCell = null;
+
 			base.List.Clear ();
 			
 			// When we clear the column collection, all rows get deleted
