--- a/mcs/class/corlib/System/Type.cs
+++ b/mcs/class/corlib/System/Type.cs
@@ -1181,18 +1181,11 @@
 				}
 			}
 			if ((memberType & MemberTypes.Property) != 0) {
-				PropertyInfo[] c;
-				int count = l.Count;
-				Type ptype;
+				PropertyInfo[] c = GetProperties (bindingAttr);
 				if (filter != null) {
-					ptype = this;
-					while ((l.Count == count) && (ptype != null)) {
-						c = ptype.GetProperties (bindingAttr);
-						foreach (MemberInfo m in c) {
-							if (filter (m, filterCriteria))
-								l.Add (m);
-						}
-						ptype = ptype.BaseType;
+					foreach (MemberInfo m in c) {
+						if (filter (m, filterCriteria))
+							l.Add (m);
 					}
 				} else {
 					c = GetProperties (bindingAttr);
--- a/mcs/class/corlib/Test/System/TypeTest.cs
+++ b/mcs/class/corlib/Test/System/TypeTest.cs
@@ -3218,6 +3218,14 @@
 				++id;
 			}
 		}
+		
+		[Test] // Bug #574696
+		public void GetMember_DoesntReturnPrivatePropOfParent ()
+		{
+			Assert.AreEqual (1, typeof (Bar).GetMember ("PrivInst", BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance).Length);
+			Assert.AreEqual (0, typeof (Bar).GetMember ("PrivInstBase", BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance).Length);
+			Assert.AreEqual (1, typeof (Foo).GetMember ("PrivInstBase", BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance).Length);
+		}
 
 #if NET_4_0
 		interface IGetInterfaceMap<in T>
