--- a/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs
+++ b/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Control.cs
@@ -2001,11 +2001,6 @@
 				return;
 			}
 
-			Point pt = PointToClient (Cursor.Position);
-
-			if (!bounds.Contains (pt) && !Capture)
-				return;
-
 			if (cursor != null || use_wait_cursor) {
 				XplatUI.SetCursor (window.Handle, Cursor.handle);
 			} else {
