--- a/mcs/class/Managed.Windows.Forms/System.Windows.Forms.Theming/Default/TabControlPainter.cs
+++ b/mcs/class/Managed.Windows.Forms/System.Windows.Forms.Theming/Default/TabControlPainter.cs
@@ -189,7 +189,7 @@
 
 			defaultFormatting = new StringFormat();
 			defaultFormatting.Alignment = StringAlignment.Near;
-			defaultFormatting.LineAlignment = StringAlignment.Near;
+			defaultFormatting.LineAlignment = StringAlignment.Center;
 			defaultFormatting.FormatFlags = StringFormatFlags.NoWrap | StringFormatFlags.NoClip;
 			defaultFormatting.HotkeyPrefix = HotkeyPrefix.Show;
 
@@ -261,6 +261,10 @@
 
 		public virtual void Draw (Graphics dc, Rectangle area, TabControl tab)
 		{
+			if (tab.SizeMode == TabSizeMode.Fixed)
+				defaultFormatting.Alignment = StringAlignment.Center;
+			else
+				defaultFormatting.Alignment = StringAlignment.Near;
 			DrawBackground (dc, area, tab);
 
 			int start = 0;
