Index: mcs/class/Managed.Windows.Forms/System.Windows.Forms/ThemeWin32Classic.cs
===================================================================
--- mcs/class/Managed.Windows.Forms/System.Windows.Forms/ThemeWin32Classic.cs	(revision 159231)
+++ mcs/class/Managed.Windows.Forms/System.Windows.Forms/ThemeWin32Classic.cs	(working copy)
@@ -2775,6 +2775,7 @@
 
 						if (image_index > -1 && image_index < control.SmallImageList.Images.Count) {
 							int image_width = control.SmallImageList.ImageSize.Width + 5;
+							int image_height = control.SmallImageList.ImageSize.Height;
 							int text_width = (int)dc.MeasureString (col.Text, control.Font).Width;
 							int x_origin = rect.X;
 
@@ -2792,7 +2793,8 @@
 							if (x_origin < rect.X)
 								x_origin = rect.X;
 
-							control.SmallImageList.Draw (dc, new Point (x_origin, rect.Y), image_index);
+							Point image_pos = new Point (x_origin, rect.Y + (rect.Height - image_height) / 2);
+							control.SmallImageList.Draw (dc, image_pos, image_index);
 							rect.X += image_width;
 							rect.Width -= image_width;
 						}
