Index: mono/mono/metadata/cominterop.c
===================================================================
--- mono/mono/metadata/cominterop.c	(revision 138231)
+++ mono/mono/metadata/cominterop.c	(working copy)   
@@ -1813,7 +1863,9 @@
 	if (!ccw_interface_hash)
 		ccw_interface_hash = g_hash_table_new (mono_aligned_addr_hash, NULL);
 
+	mono_cominterop_lock();
 	ccw_list = g_hash_table_lookup (ccw_hash, GINT_TO_POINTER (mono_object_hash (object)));
+	mono_cominterop_unlock();
 
 	ccw_list_item = ccw_list;
 	while (ccw_list_item) {
