Index: mcs/class/System.XML/System.Xml/XmlNode.cs
===================================================================
--- mcs/class/System.XML/System.Xml/XmlNode.cs	(revision 157243)
+++ mcs/class/System.XML/System.Xml/XmlNode.cs	(revision 157244)
@@ -864,7 +864,7 @@
 			XPathNodeIterator iter = nav.Select (expr);
 			if (!iter.MoveNext ())
 				return null;
-			return ((IHasXmlNode) iter.Current).GetNode ();
+			return (iter.Current as IHasXmlNode).GetNode ();
 		}
 
 		public virtual bool Supports (string feature, string version)
