Index: mcs/class/System.XML/System.Xml/XmlNode.cs
===================================================================
--- mcs/class/System.XML/System.Xml/XmlNode.cs	(revision 152577)
+++ mcs/class/System.XML/System.Xml/XmlNode.cs	(working copy)
@@ -39,6 +39,7 @@
 #if NET_2_0
 using System.Diagnostics;
 using System.Xml.Schema;
+using System.Collections.Generic;
 #endif
 
 namespace System.Xml
@@ -855,16 +856,27 @@
 			return SelectSingleNode (xpath, null);
 		}
 
+		static Dictionary<string, XmlNode> m_cache = new Dictionary<string, XmlNode>();
+
 		public XmlNode SelectSingleNode (string xpath, XmlNamespaceManager nsmgr)
 		{
-			XPathNavigator nav = CreateNavigator ();
-			XPathExpression expr = nav.Compile (xpath);
-			if (nsmgr != null)
-				expr.SetContext (nsmgr);
-			XPathNodeIterator iter = nav.Select (expr);
-			if (!iter.MoveNext ())
-				return null;
-			return ((IHasXmlNode) iter.Current).GetNode ();
+			XmlNode ret;
+			if (!m_cache.TryGetValue(xpath + this.InnerXml, out ret))
+ 			{
+ 				XPathNavigator nav = CreateNavigator ();
+ 				XPathExpression expr = nav.Compile (xpath);
+ 				if (nsmgr != null)
+ 					expr.SetContext (nsmgr);
+ 				XPathNodeIterator iter = nav.Select (expr);
+ 				if (!iter.MoveNext ())
+ 					ret = null;
+ 				else
+ 					ret = (iter.Current as IHasXmlNode).GetNode ();
+ 
+ 				m_cache.Add(xpath + this.InnerXml, ret);
+ 			}
+ 
+ 			return ret;
 		}
 
 		public virtual bool Supports (string feature, string version)
