FWNX-450 fix backuprestore tests

Index: fieldworks-mono/mcs/class/System.Runtime.Serialization/System.Runtime.Serialization/SerializationMap.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/System.Runtime.Serialization/System.Runtime.Serialization/SerializationMap.cs
+++ fieldworks-mono/mcs/class/System.Runtime.Serialization/System.Runtime.Serialization/SerializationMap.cs
@@ -420,6 +420,7 @@
 
 				if (dmi == null) {
 					reader.Skip ();
+					reader.MoveToContent ();
 					continue;
 				}
 				SetValue (dmi, instance, deserializer.Deserialize (dmi.MemberType, reader));
Index: fieldworks-mono/mcs/class/System.Runtime.Serialization/Test/System.Runtime.Serialization/XmlObjectSerializerTest.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/System.Runtime.Serialization/Test/System.Runtime.Serialization/XmlObjectSerializerTest.cs
+++ fieldworks-mono/mcs/class/System.Runtime.Serialization/Test/System.Runtime.Serialization/XmlObjectSerializerTest.cs
@@ -1478,6 +1478,19 @@
 			
 			var ds = (DataSet) x.ReadObject (r);
 		}
+
+		[Test]
+		public void Bug642227_IgnoreUnknownElement()
+		{
+			string xml = @"
+<DataContractSerializerTest.Sample1 xmlns=""http://schemas.datacontract.org/2004/07/MonoTests.System.Runtime.Serialization"">
+	<Member2>bla</Member2>
+	<Member1>bar</Member1>
+</DataContractSerializerTest.Sample1>";
+			new DataContractSerializer (typeof (Sample1))
+				.ReadObject (XmlReader.Create (new StringReader (xml)));
+		}
+
 	}
 	
 	[DataContract]
