Index: fieldworks-mono/mcs/class/corlib/System.Resources/ResourceManager.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/corlib/System.Resources/ResourceManager.cs
+++ fieldworks-mono/mcs/class/corlib/System.Resources/ResourceManager.cs
@@ -334,4 +334,6 @@
  			
 			/* if we already have this resource set, return it */
+#if false
 			set = (ResourceSet) ResourceSets [culture];
+#endif
 			if (set != null) {
