FWNX-597

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStrip.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStrip.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStrip.cs
@@ -1329,19 +1329,23 @@
 		{
 			get { return this.keyboard_active; }
 			set {
-				if (this.keyboard_active != value) {
-					this.keyboard_active = value;
-					
-					if (value)
-						Application.KeyboardCapture = this;
-					else if (Application.KeyboardCapture == this) {
+				if (keyboard_active == value == true)
+					return;
+
+				if (keyboard_active == value == false && Application.KeyboardCapture == null)
+					return;
+
+				keyboard_active = value;
+
+				if (value)
+					Application.KeyboardCapture = this;
+				else if (Application.KeyboardCapture == this || Items.Contains (Application.KeyboardCapture)) {
 						Application.KeyboardCapture = null;
 						ToolStripManager.ActivatedByKeyboard = false;
-					}
-					
-					// Redraw for mnemonic underlines
-					this.Invalidate ();
 				}
+
+				// Redraw for mnemonic underlines
+				this.Invalidate ();		
 			}
 		}
 		#endregion
