FWNX-340 and FWNX-160

Index: mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs
===================================================================
--- mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs.orig	2010-11-09 22:08:15.000000000 -0700
+++ mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/DataGridView.cs	2010-11-09 22:08:52.000000000 -0700
@@ -6272,7 +6272,7 @@
 					if (disp_x == Columns.Count - 1)
 						delta_x = horizontalScrollBar.Maximum - horizontalScrollBar.Value;
 					else
-						for (int i = first_col_index + displayedColumnsCount - 1; i < disp_x; i++)
+						for (int i = first_col_index + displayedColumnsCount - 1; i < disp_x && i != -1; i++)
 							delta_x += Columns[ColumnDisplayIndexToIndex (i)].Width;
 
 					horizontalScrollBar.SafeValueSet (horizontalScrollBar.Value + delta_x);
