Index: fieldworks-mono/mcs/class/System.ServiceModel/System.IO/PipeException.cs
===================================================================
--- /dev/null
+++ fieldworks-mono/mcs/class/System.ServiceModel/System.IO/PipeException.cs
@@ -0,0 +1,39 @@
+using System;
+using System.Runtime.Serialization;
+using System.ServiceModel.Channels;
+
+namespace System.IO
+{
+	public class PipeException : IOException
+	{
+		public PipeException()
+		{
+		}
+
+		public PipeException(string message): base(message)
+		{
+		}
+
+		protected PipeException(SerializationInfo info, StreamingContext context): base(info, context)
+		{
+		}
+
+		public PipeException(string message, Exception inner): base(message, inner)
+		{
+		}
+
+		public PipeException(string message, int errorCode): base(message, errorCode)
+		{
+		}
+
+		public virtual int ErrorCode
+		{
+			get
+			{
+				// we re-use the HResult for the error code here.
+				return HResult;
+			}
+		}
+	}
+}
+
Index: fieldworks-mono/mcs/class/System.ServiceModel/System.ServiceModel.dll.sources
===================================================================
--- fieldworks-mono.orig/mcs/class/System.ServiceModel/System.ServiceModel.dll.sources
+++ fieldworks-mono/mcs/class/System.ServiceModel/System.ServiceModel.dll.sources
@@ -128,6 +128,7 @@
 System.Collections.Generic/SynchronizedCollection.cs
 System.Collections.Generic/SynchronizedKeyedCollection.cs
 System.Collections.Generic/SynchronizedReadOnlyCollection.cs
+System.IO/PipeException.cs
 System.ServiceModel.Activation/AspNetIntegrationRequirementsAttribute.cs
 System.ServiceModel.Activation/IServiceHostFactory.cs
 System.ServiceModel.Activation/ServiceHostFactory.cs
Index: fieldworks-mono/mcs/class/System.ServiceModel/System.ServiceModel/Dummy.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/System.ServiceModel/System.ServiceModel/Dummy.cs
+++ fieldworks-mono/mcs/class/System.ServiceModel/System.ServiceModel/Dummy.cs
@@ -1,10 +1,5 @@
 using System.ServiceModel.Channels;
 
-namespace System.IO
-{
-	public class PipeException : Exception { }
-}
-
 namespace System.ServiceModel
 {
 	public class PeerHopCountAttribute { }
