FWNX-272

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
@@ -449,20 +449,20 @@
 							direction = ToolStripDropDownDirection.BelowRight;
 						break;
 					case ToolStripDropDownDirection.BelowLeft:
-						if (show_point.Y + this.Height > max_screen.Y)
+						if (show_point.Y + this.Height > max_screen.Y && show_point.Y - this.Height > 0)
 							direction = ToolStripDropDownDirection.AboveLeft;
 						break;
 					case ToolStripDropDownDirection.BelowRight:
 					case ToolStripDropDownDirection.Default:
-						if (show_point.Y + this.Height > max_screen.Y)
+						if (show_point.Y + this.Height > max_screen.Y && show_point.Y - this.Height > 0)
 							direction = ToolStripDropDownDirection.AboveRight;
 						break;
 					case ToolStripDropDownDirection.Left:
-						if (show_point.Y + this.Height > max_screen.Y)
+						if (show_point.Y + this.Height > max_screen.Y && show_point.Y - this.Height > 0)
 							direction = ToolStripDropDownDirection.AboveLeft;
 						break;
 					case ToolStripDropDownDirection.Right:
-						if (show_point.Y + this.Height > max_screen.Y)
+						if (show_point.Y + this.Height > max_screen.Y && show_point.Y - this.Height > 0)
 							direction = ToolStripDropDownDirection.AboveRight;
 						break;
 				}
@@ -486,12 +486,18 @@
 					break;
 			}
 
-			// Fix offscreen positions
+			// Fix offscreen horizontal positions
 			if ((show_point.X + this.Width) > max_screen.X)
 				show_point.X = max_screen.X - this.Width;
 			if (show_point.X < 0)
 				show_point.X = 0;
-			
+
+			// Fix offscreen vertical positions
+			if ((show_point.Y + this.Height) > max_screen.Y)
+				show_point.Y = max_screen.Y - this.Height;
+			if (show_point.Y < 0)
+				show_point.Y = 0;
+
 			if (this.Location != show_point)
 				this.Location = show_point;
 
