FWNX-434

Index: fieldworks-mono/mcs/class/corlib/System/__ComObject.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/corlib/System/__ComObject.cs
+++ fieldworks-mono/mcs/class/corlib/System/__ComObject.cs
@@ -39,6 +39,33 @@
 
 namespace System
 {
+	/// <summary>
+	/// Interface to allow alternative ComObject release mechanisms to be specified.
+	/// </summary>
+	public interface IComObjectReleaser
+	{
+		/// <summary>
+		/// Marshal.Release needs to be called on pUnk parameter, to prevent memory leaks.
+		/// </summary>		
+		void Add(IntPtr pUnk);
+	}
+	
+	/// <summary>
+	/// Register an external IComObjectReleaser.
+	/// </summary>
+	public static class __ComObjectReleaser
+	{	
+		internal static IComObjectReleaser ComObjectReleaserImplementation;
+		
+		/// <summary>
+		/// Supply a alternative ComObject release implementation.		
+		/// </summary>
+		public static void Register(IComObjectReleaser comObjectReleaser)
+		{
+			ComObjectReleaserImplementation = comObjectReleaser;
+		}					
+	}
+	
 	// This is a private class that is used as a generic wrapper class
 	// for COM objects that have no specific wrapper class.
 	//
@@ -49,7 +76,7 @@
 	// This class is referenced in .NET Framework SDK Documentation so
 	// many times that obj.GetType().FullName == "System.__ComObject" and
 	// Type.GetType("System.__ComObject") may be used.
-
+	
 	internal class __ComObject : MarshalByRefObject
 	{
 #pragma warning disable 169	
@@ -66,7 +93,13 @@
 		private extern void ReleaseInterfaces ();
 
 		~__ComObject ()
-		{
+		{	
+			// If external ComObjectReleaserImplementation isn't supplied ComObject gets released by the GC thread.			
+			if (__ComObjectReleaser.ComObjectReleaserImplementation != null && iunknown != IntPtr.Zero)
+			{
+				Marshal.AddRef(iunknown);
+				__ComObjectReleaser.ComObjectReleaserImplementation.Add(iunknown);			
+			}
 			ReleaseInterfaces ();
 		}
 
