FWNX-434 +GeckoFx

Index: fieldworks-mono/mcs/class/corlib/System/__ComObject.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/corlib/System/__ComObject.cs
+++ fieldworks-mono/mcs/class/corlib/System/__ComObject.cs
@@ -36,6 +36,7 @@
 using System.Collections;
 using System.Runtime.InteropServices;
 using System.Runtime.CompilerServices;
+using System.Threading;
 
 namespace System
 {
@@ -56,6 +57,7 @@
 		#region Sync with object-internals.h
 		IntPtr iunknown;
 		IntPtr hash_table;
+		SynchronizationContext synchronization_context;
 		#endregion
 #pragma warning restore 169
 
@@ -66,8 +68,11 @@
 		private extern void ReleaseInterfaces ();
 
 		~__ComObject ()
-		{
-			ReleaseInterfaces ();
+		{	
+			if (synchronization_context != null)
+				synchronization_context.Post ((state) => ReleaseInterfaces (), this);
+			else
+				ReleaseInterfaces ();				
 		}
 
 		public __ComObject ()
@@ -81,6 +86,7 @@
 
 		internal __ComObject (IntPtr pItf)
 		{
+			InitializeApartmentDetails ();
 			Guid iid = IID_IUnknown;
 			int hr = Marshal.QueryInterface (pItf, ref iid, out iunknown);
 			Marshal.ThrowExceptionForHR (hr);
@@ -88,6 +94,7 @@
 
 		internal void Initialize (Type t)
 		{
+			InitializeApartmentDetails ();
 			// Guard multiple invocation.
 			if (iunknown != IntPtr.Zero)
 				return;
@@ -104,6 +111,21 @@
 			}
 		}
 
+		private void InitializeApartmentDetails ()
+		{
+			// Only synchronization_context if thread is STA.
+			if (Thread.CurrentThread.GetApartmentState() != ApartmentState.STA)
+				return;
+			
+			synchronization_context = SynchronizationContext.Current;
+
+			// Check whether the current context is a plain SynchronizationContext object
+			// and handle this as if no context was set at all.
+			if (synchronization_context != null &&
+				synchronization_context.GetType () == typeof(SynchronizationContext))
+				synchronization_context = null;			
+		}
+
 		private static Guid GetCLSID (Type t)
 		{
 			if (t.IsImport)
Index: fieldworks-mono/mono/metadata/object-internals.h
===================================================================
--- fieldworks-mono.orig/mono/metadata/object-internals.h
+++ fieldworks-mono/mono/metadata/object-internals.h
@@ -322,6 +322,7 @@
 	MonoMarshalByRefObject object;
 	gpointer iunknown;
 	GHashTable* itf_hash;
+	MonoObject *synchronization_context;
 } MonoComObject;
 
 typedef struct {
