FWNX-512 fix modal form cancel bug

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Form.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Form.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Form.cs
@@ -712,7 +712,8 @@
 							typeof (DialogResult));
 
 				dialog_result = value;
-				closing = (dialog_result != DialogResult.None && is_modal);
+				if (dialog_result != DialogResult.None && is_modal)
+					RaiseCloseEvents(false, false);
 			}
 		}
 
