FWNX-597

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/ToolStripDropDown.cs
@@ -934,6 +934,14 @@
 		{
 			this.Close (reason);
 			base.Dismiss (reason);
+
+			// ContextMenuStrip won't have a parent
+			if (this.OwnerItem == null)
+				return;
+			
+			// Ensure Submenu loses keyboard capture when closing.
+			ToolStrip parent_strip = this.OwnerItem.Parent;
+			ToolStripManager.SetActiveToolStrip (null, false);			
 		}
 
 		internal override ToolStrip GetTopLevelToolStrip ()
