FWNX-439 problems with the backspace key being flaky

Index: mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/X11Keyboard.cs
===================================================================
--- mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/X11Keyboard.cs.orig	2010-11-09 22:08:53.000000000 -0700
+++ mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/X11Keyboard.cs	2010-11-09 22:08:54.000000000 -0700
@@ -52,7 +52,7 @@
 		private XIMProperties ximStyle;
 		private EventMask xic_event_mask = EventMask.NoEventMask;
 		private StringBuilder lookup_buffer;
-		private byte [] utf8_buffer;
+		private byte [] lookup_byte_buffer = new byte [100];
 		private int min_keycode, max_keycode, keysyms_per_keycode, syms;
 		private int [] keyc2vkey = new int [256];
 		private int [] keyc2scan = new int [256];
@@ -131,8 +131,6 @@
 			xim = XOpenIM (display, IntPtr.Zero, IntPtr.Zero, IntPtr.Zero);
 			if (xim == IntPtr.Zero) 
 				Console.Error.WriteLine ("Could not get XIM");
-			else 
-				utf8_buffer = new byte [100];
 
 			initialized = true;
 		}
@@ -1184,19 +1182,26 @@
 			IntPtr xic = GetXic (client_window);
 			if (xic != IntPtr.Zero) {
 				do {
-					res = Xutf8LookupString (xic, ref xevent, utf8_buffer, 100, out keysym_res,  out status);
+					res = Xutf8LookupString (xic, ref xevent, lookup_byte_buffer, 100, out keysym_res,  out status);
 					if ((int) status != -1) // XLookupBufferOverflow
 						break;
-					utf8_buffer = new byte [utf8_buffer.Length << 1];
+					lookup_byte_buffer = new byte [lookup_byte_buffer.Length << 1];
 				} while (true);
 				lookup_buffer.Length = 0;
-				string s = Encoding.UTF8.GetString (utf8_buffer, 0, res);
+				string s = Encoding.UTF8.GetString (lookup_byte_buffer, 0, res);
 				lookup_buffer.Append (s);
 				keysym = (XKeySym) keysym_res.ToInt32 ();
 				return s.Length;
 			} else {
+				do {
+					res = XLookupString (ref xevent, lookup_byte_buffer, 100, out keysym_res, out status);
+					if ((int) status != -1) // XLookupBufferOverflow
+						break;
+					lookup_byte_buffer = new byte [lookup_byte_buffer.Length << 1];
+				} while (true);
 				lookup_buffer.Length = 0;
-				res = XLookupString (ref xevent, lookup_buffer, len, out keysym_res, IntPtr.Zero);
+				string s = Encoding.ASCII.GetString (lookup_byte_buffer, 0, res);
+				lookup_buffer.Append (s);
 				keysym = (XKeySym) keysym_res.ToInt32 ();
 				return res;
 			}
@@ -1262,7 +1267,7 @@
 		private static extern bool XSetLocaleModifiers (string mods);
 
 		[DllImport ("libX11")]
-		internal extern static int XLookupString(ref XEvent xevent, StringBuilder buffer, int num_bytes, out IntPtr keysym, IntPtr status);
+		internal extern static int XLookupString(ref XEvent xevent, byte [] buffer, int num_bytes, out IntPtr keysym, out IntPtr status);
 		[DllImport ("libX11")]
 		internal extern static int Xutf8LookupString(IntPtr xic, ref XEvent xevent, byte [] buffer, int num_bytes, out IntPtr keysym, out IntPtr status);
 
