Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms.VisualStyles/VisualStylesGtkPlus.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms.VisualStyles/VisualStylesGtkPlus.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms.VisualStyles/VisualStylesGtkPlus.cs
@@ -1100,7 +1100,7 @@
 		}
 		public string VisualStyleInformationColorScheme {
 			get {
-				return null;
+				return Environment.GetEnvironmentVariable ("MONO_VISUAL_STYLE_COLOR_SCHEME");
 			}
 		}
 		public string VisualStyleInformationCompany {
@@ -1130,7 +1130,7 @@
 		}
 		public string VisualStyleInformationFileName {
 			get {
-				return null;
+				return VisualStyleInformationColorScheme;
 			}
 		}
 		public bool VisualStyleInformationIsSupportedByOS
