diff --git a/mcs/class/System.ServiceModel/System.ServiceModel/Dummy.cs b/mcs/class/System.ServiceModel/System.ServiceModel/Dummy.cs
index 5a1241d..9c1e85a 100644
--- a/mcs/class/System.ServiceModel/System.ServiceModel/Dummy.cs
+++ b/mcs/class/System.ServiceModel/System.ServiceModel/Dummy.cs
@@ -1,10 +1,5 @@
 using System.ServiceModel.Channels;
 
-namespace System.IO
-{
-	public class PipeException : Exception { }
-}
-
 namespace System.ServiceModel
 {
 	public class PeerHopCountAttribute { }
diff --git a/mcs/class/System.ServiceModel/System.ServiceModel/PipeException.cs b/mcs/class/System.ServiceModel/System.ServiceModel/PipeException.cs
new file mode 100644
index 0000000..65c155c
--- /dev/null
+++ b/mcs/class/System.ServiceModel/System.ServiceModel/PipeException.cs
@@ -0,0 +1,37 @@
+using System.ServiceModel.Channels;
+
+namespace System.IO
+{
+	public class PipeException : IOException
+	{
+		public PipeException()
+		{
+		}
+
+		public PipeException(string message): base(message)
+		{
+		}
+
+		protected PipeException(SerializationInfo info, StreamingContext context): base(info, context)
+		{
+		}
+
+		public PipeException(string message, Exception inner): base(message, inner)
+		{
+		}
+
+		public PipeException(string message, int errorCode): base(message, errorCode)
+		{
+		}
+
+		public virtual int ErrorCode
+		{
+			get
+			{
+				// we re-use the HResult for the error code here.
+				return HResult;
+			}
+		}
+	}
+}
+
