From da0c0605fb25950173b93c735b24477cd9daf778 Mon Sep 17 00:00:00 2001
From: Quilt patch <unknown@unknown.com>
Date: Tue, 30 Aug 2011 16:56:38 +0200
Subject: [PATCH] WORK_AROUND_BAD_NATIVE_CODE

---
 mono/metadata/cominterop.c |   13 +++++++++++++
 1 files changed, 13 insertions(+), 0 deletions(-)

Index: fieldworks-mono/mono/metadata/cominterop.c
===================================================================
--- fieldworks-mono.orig/mono/metadata/cominterop.c
+++ fieldworks-mono/mono/metadata/cominterop.c
@@ -34,6 +34,7 @@
 #include "mono/utils/mono-counters.h"
 #include <string.h>
 #include <errno.h>
+#include <signal.h>
 
 #ifndef DISABLE_COM
 
@@ -1664,6 +1665,11 @@
 	return TRUE;
 }
 
+void justexit(int sig)
+{
+	_exit(0);
+}
+
 void
 cominterop_release_all_rcws (void)
 {
@@ -1671,11 +1677,18 @@
 		return;
 
 	mono_cominterop_lock ();
+	sighandler_t oldAbrt = signal(SIGABRT, justexit);
+	sighandler_t oldSegv = signal(SIGSEGV, justexit);
+	sighandler_t oldFpe = signal(SIGFPE, justexit);
 
 	g_hash_table_foreach_remove (rcw_hash, cominterop_rcw_finalizer, NULL);
 	g_hash_table_destroy (rcw_hash);
 	rcw_hash = NULL;
 
+	signal(SIGABRT, oldAbrt);
+	signal(SIGABRT, oldSegv);
+	signal(SIGFPE, oldFpe);
+
 	mono_cominterop_unlock ();
 }
 
